### PR TITLE
Optimize JthetaTmv

### PR DESCRIPTION
**Pull request by klensink**
- Type Stabilize
- Dot Vectorize
- View sub arrays to prevent re-allocating

```
julia> judge(hist)
BenchmarkTools.TrialJudgement: 
  time:   -61.17% => improvement (5.00% tolerance)
  memory: +2.63% => regression (1.00% tolerance)
```

### CODE REVIEW

File: src/kernelTypes/sparseKernel.jl [hunk 1]
```diff
@@ -68,17 +68,30 @@ function Jthetamv(this::SparseKernel,dtheta,theta,Y,tmp=nothing)
...
+        @simd for i=1:length(this.ival)
             @inbounds t[i] += Z[this.ival[i],j] * Y[this.jval[i],j]
         end
     end
 
-    return this.Qs'*t;
+    return this.Qs'*t
+end
+
+function JthetaTmv{T}(this::SparseKernel{T}, Z_in, theta, Y_in, tmp=nothing)
```
Comment (klensink) [inline src/kernelTypes/sparseKernel.jl:86]: @eldadHaber @lruthotto 

Can you confirm this, and the change to line 13 are okay? 

I've checked that all tests pass and that `JthetaTmv` still produces the same output as before, I just want to make sure

Comment (lruthotto) [reply src/kernelTypes/sparseKernel.jl:86]: If you mean the type of Qs that could be a problem. As far as I know, `SparseMatrixCSC` is not a subtype of `Array`, is it? There will be applications where `Qs` is sparse. Please double check.

**Comment by eldadHaber** [comment]
This looks perfect to me

**Review by DavidBegert — APPROVED** [review]

File: src/kernelTypes/sparseKernel.jl (hunk 1 above)
Comment (klensink) [reply src/kernelTypes/sparseKernel.jl:86]: You were right, I removed that change

**Pull request merged**